### DISCUSSION TITLE
"Property 'getTeams' is incompatible with index signature." error with npm run build

### DISCUSSION THREAD

**Discussion by FlyingTurkman** [vercel/next.js#56376] [Help]
### Summary

I want to use server side generation in a page with an api but when I tried to build giving this error
Type 'OmitWithTag<typeof import("C:/Users/deniz/OneDrive/Masa\u00FCst\u00FC/psoleaguev2/psoleaguev2/app/teams/page"), "default" | "metadata" | "config" | "generateStaticParams" | "revalidate" | "dynamic" | ... 5 more ... | "generateMetadata", "">' does not satisfy the constraint '{ [x: string]: never; }'.
  Property 'getTeams' is incompatible with index signature.
    Type '() => Promise<teamType[]>' is not assignable to type 'never'.

Page codes below:

```tsx
export default async function Page() {
    const teams: teamType[] = await getTeams()
    return(
        <div className="flex flex-col w-full items-center">
            <PageView initialTeams={teams}/>
        </div>
    )
}


export async function getTeams(): Promise<teamType[]> {
    try {
        const resTeams = await fetch(`${process.env.appPath}/api/teamApi/getLastTeamsApi`, {
            method: 'POST',
            body: JSON.stringify({
                apiSecret: process.env.apiSecret
            }),
            cache: 'no-cache'
        })

        const res = await resTeams.json()

        if (resTeams.status == 200) {
            return res
        } else {
            return []
        }
    } catch (error) {
        console.log(error)
        return []
    }
}

```

### Additional information

```js
Next js 13.5.3
I don't want to use "ignoreBuildErrors: true" at next.config.js
```

**Comment by icyJoseph** [comment]
I can't see the TS error: https://stackblitz.com/edit/github-bff8j8?file=app%2Flayout.tsx,app%2Fteams%2Fpage.tsx

But I can already tell you that the way you are doing it, is not going to work. Please refer to this FAQ:

- https://nextjs-faq.com/fetch-api-in-rsc

You can't fetch from API endpoint during build time, and even if you want to force it by setting localhost:3000, to a dynamic value, it is still very inefficient, and unnecessary. You are already server side, so just fetch the data there.

**Reply by FlyingTurkman** [reply]
I tried like this but still same error.

```tsx
export async function getTeams(): Promise<teamType[]> {
    try {
        const teams: teamType[] = await Teams.find({}).limit(20)
        return teams
    } catch (error) {
        console.log(error)
        return []
    }
}
```

**Reply by icyJoseph** [reply]
Is this using MongoDB or something? Could you please paste a screenshot of the error?

**Reply by FlyingTurkman** [reply]
Yes MongoDb 
![image](https://github.com/vercel/next.js/assets/99084104/7cc43b9f-9cc2-4138-a487-dbb3388cc7b4)

**Reply by icyJoseph** [reply]
Ohhh I think I know. It is because you are exporting something that pages within App Router don't expect to be exported. This bit:

```tsx
export async function getTeams(/* rest of implementation */
```

I think that the Next.js App Router compiler is very strict with what we are allowed to export from `page.tsx` and others. I guess you could try to refactor, by moving getTeams to another file, and import it into the page from there, and wherever else it is needed from that new file as well.

**Reply by FlyingTurkman** [reply]
I will try it thank you

**Reply by euclidesdry-tripee** [reply]
> Ohhh I think I know. It is because you are exporting something that pages within App Router don't expect to be exported. This bit:
> 
> ```tsx
> export async function getTeams(/* rest of implementation */
> ```
> 
> I think that the Next.js App Router compiler is very strict with what we are allowed to export from `page.tsx` and others. I guess you could try to refactor, by moving getTeams to another file, and import it into the page from there, and wherever else it is needed from that new file as well.

**The best answer ever is that it only needs to have attention to interpret the Exception.**

**Comment by FlyingTurkman** [comment]
Btw its working at live
